client/electron/go_plugin.ts

```diff
@@ -36,7 +34,99 @@ export async function invokeGoMethod(
   method: string,
   input: string
 ): Promise<string> {
-  if (!invokeMethodFunc) {
+  console.debug(`[Backend] - calling InvokeMethod "${method}" ...`);
+  const result = await ensureCgo().invokeMethod(method, input);
+  console.debug(`[Backend] - InvokeMethod "${method}" returned`, result);
+  if (result.ErrorJson) {
+    throw new Error(result.ErrorJson);
+  }
+  return result.Output;
+}
+
+/**
+ * Represents a function that will be called from the Go backend.
+ * @param data The data string passed from the Go backend.
+ * @returns A result string that will be passed back to the caller.
+ */
+export type CallbackFunction = (data: string) => string;
+
+/** A unique token for a callback created by `newCallback`. */
+export type CallbackToken = number;
+
+/**
+ * Koffi requires us to register all persistent callbacks; we track the registrations here.
+ * @see https://koffi.dev/callbacks#registered-callbacks
+ */
+const koffiCallbacks = new Map<CallbackToken, koffi.IKoffiRegisteredCallback>();
+
+/**
+ * Registers a callback function in TypeScript, making it invokable from Go.
+ *
+ * The caller can delete the callback by calling `deleteCallback`.
+ *
+ * @param callback The callback function to be registered.
+ * @returns A Promise resolves to the callback token, which can be used to refer to the callback.
+ */
+export async function newCallback(
+  callback: CallbackFunction
+): Promise<CallbackToken> {
+  console.debug('[Backend] - calling newCallback ...');
+  const persistentCallback = koffi.register(
+    callback,
+    ensureCgo().callbackFuncPtr
+  );
+  const token = await ensureCgo().newCallback(persistentCallback);
+  console.debug('[Backend] - newCallback done, token:', token);
+  koffiCallbacks.set(token, persistentCallback);
+  return token;
+}
+
+/**
+ * Unregisters a specified callback function from the Go backend.
+ *
+ * @param token The callback token returned from `newCallback`.
+ * @returns A Promise that resolves when the unregistration is done.
+ */
+export async function deleteCallback(token: CallbackToken): Promise<void> {
+  console.debug('[Backend] - calling deleteCallback:', token);
+  await ensureCgo().deleteCallback(token);
+  console.debug('[Backend] - deleteCallback done');
+  const persistentCallback = koffiCallbacks.get(token);
+  if (persistentCallback) {
+    koffi.unregister(persistentCallback);
+    koffiCallbacks.delete(token);
+    console.debug(
+      `[Backend] - unregistered persistent callback ${token} from koffi`
+    );
+  }
+}
+
+/** Interface containing the exported native CGo functions. */
+interface CgoFunctions {
```

Comment:
The encapsulation here is a bit off. It can be more object-oriented and remove globals.

Perhaps call this CallbackMap with public New and Delete methods.
Make the fields private.
Move koffiCallbacks into here.

The logic in ensureCgo can be in the constructor or factory function. Then there's no need to call ensureCgo in the methods.

Perhaps rename ensureCgo to getCallbackMap()

This makes the code more encapsulated, safer and with less globals.

Reply:
It involves multiple components, I would prefer `getBackendLib`, and returns an object with a `methodChannel` and a `callbackManager` object.